File: tools/sentinel/src/lib/tools/audio-refinery.ts
/**
 * 🎧 Audio Refinery — Convert Markdown files to MP3 via Edge TTS
 * Replaces tools/audio_refinery/script.py
 *
 * Usage (direct): tsx src/lib/tools/audio-refinery.ts <file_or_dir> [voice]
 */

import { readFile, writeFile, mkdir, stat, readdir } from 'fs/promises';
import { existsSync } from 'fs';
import { join, dirname, basename, extname } from 'path';
import { EdgeTTS } from '@travisvn/edge-tts';

const DEFAULT_VOICE = 'zh-CN-XiaoxiaoNeural';
const CONCURRENCY = 3;

function cleanMarkdown(text: string): string {
    return text
        .replace(/^---[\s\S]*?---/m, '')           // YAML frontmatter
        .replace(/```[\s\S]*?```/g, '')             // code blocks
        .replace(/`[^`]*`/g, '')                    // inline code
        .replace(/^#{1,6}\s/gm, '')                 // headings
        .replace(/\*\*|__|[*_]/g, '')               // bold / italic
        .replace(/!\[([^\]]*)\]\([^)]+\)/g, '')     // images
        .replace(/\[([^\]]+)\]\([^)]+\)/g, '$1')   // links → text
        .replace(/\[\[(?:[^|\]]*\|)?([^\]]+)\]\]/g, '$1') // WikiLinks
        .replace(/!\[\[.*?\]\]/g, '')               // Obsidian embeds
        .replace(/<[^>]+>/g, '')                    // HTML tags
        .replace(/\n{3,}/g, '\n\n')
        .trim();
}

async function processFile(filePath: string, voice: string): Promise<string> {
    const dir = dirname(filePath);
    const name = basename(filePath, extname(filePath));
    const audioDir = join(dir, 'Audio');
    const outputPath = join(audioDir, `${name}.mp3`);

    await mkdir(audioDir, { recursive: true });

    // Incremental: skip if mp3 is newer than md
    if (existsSync(outputPath)) {
        const [mdStat, mp3Stat] = await Promise.all([stat(filePath), stat(outputPath)]);
        if (mdStat.mtimeMs < mp3Stat.mtimeMs) {
            return `⏩ Skipped (up to date): ${basename(filePath)}`;
        }
    }

    const raw = await readFile(filePath, 'utf-8');
    const cleanText = cleanMarkdown(raw);
    if (!cleanText) return `⚠️ Skipped (empty): ${basename(filePath)}`;

    const tts = new EdgeTTS(cleanText, voice);
    const { audio } = await tts.synthesize();
    const buf = Buffer.from(await audio.arrayBuffer());
    await writeFile(outputPath, buf);

    // Back-fill embed link into Markdown
    const embedString = `![[Audio/${name}.mp3]]`;
    if (!raw.includes(embedString)) {
        const yamlMatch = raw.match(/^---[\s\S]*?---\n/m);
        const newContent = yamlMatch
            ? raw.slice(0, yamlMatch[0].length) + `\n${embedString}\n` + raw.slice(yamlMatch[0].length)
            : `${embedString}\n\n${raw}`;
        await writeFile(filePath, newContent, 'utf-8');
    }

    return `✅ Generated: ${basename(outputPath)}`;
}

async function runWithConcurrency<T>(tasks: (() => Promise<T>)[], limit: number): Promise<T[]> {
    const results: T[] = [];
    const queue = [...tasks];
    const workers = Array(Math.min(limit, tasks.length)).fill(null).map(async () => {
        while (queue.length > 0) {
            const task = queue.shift()!;
            results.push(await task());
        }
    });
    await Promise.all(workers);
    return results;
}

export async function audioRefinery(target: string, voice = DEFAULT_VOICE): Promise<string> {
    const files: string[] = [];

    const s = await stat(target).catch(() => null);
    if (!s) throw new Error(`Path not found: ${target}`);

    if (s.isFile() && (target.endsWith('.md') || target.endsWith('.txt'))) {
        files.push(target);
    } else if (s.isDirectory()) {
        const walk = async (dir: string) => {
            const entries = await readdir(dir, { withFileTypes: true });
            for (const e of entries) {
                const full = join(dir, e.name);
                if (e.isDirectory() && e.name !== 'Audio') await walk(full);
                else if (e.isFile() && e.name.endsWith('.md')) files.push(full);
            }
        };
        await walk(target);
    }

    if (files.length === 0) return '⚠️ No Markdown files found.';

    const tasks = files.map(f => () => processFile(f, voice));
    const results = await runWithConcurrency(tasks, CONCURRENCY);

    return results.join('\n');
}

// Run directly
if (import.meta.url === `file://${process.argv[1]}`) {
    const target = process.argv[2];
    if (!target) { console.error('Usage: tsx audio-refinery.ts <file_or_dir> [voice]'); process.exit(1); }
    console.log(`🎧 Audio Refinery starting...`);
    const output = await audioRefinery(target, process.argv[3]);
    console.log(output);
}

File: app/refinery/audio_refinery.py
import os
import sys
import asyncio
from pathlib import Path
import re
import edge_tts

DEFAULT_VOICE = 'zh-CN-XiaoxiaoNeural'
CONCURRENCY = 3

def clean_markdown(text: str) -> str:
    text = re.sub(r'^---[\s\S]*?---\n', '', text, flags=re.MULTILINE) # frontmatter
    text = re.sub(r'```[\s\S]*?```', '', text) # code blocks
    text = re.sub(r'`[^`]*`', '', text) # inline code
    text = re.sub(r'^#{1,6}\s', '', text, flags=re.MULTILINE) # headings
    text = re.sub(r'\*\*|__|[*_]', '', text) # bold/italic
    text = re.sub(r'!\[([^\]]*)\]\([^)]+\)', '', text) # images
    text = re.sub(r'\[([^\]]+)\]\([^)]+\)', r'\1', text) # links -> text
    text = re.sub(r'\[\[(?:[^|\]]*\|)?([^\]]+)\]\]', r'\1', text) # WikiLinks
    text = re.sub(r'!\[\[.*?\]\]', '', text) # Obsidian embeds
    text = re.sub(r'<[^>]+>', '', text) # HTML tags
    text = re.sub(r'\n{3,}', '\n\n', text) # excessive newlines
    return text.strip()

async def process_file(file_path: Path, voice: str) -> str:
    name = file_path.stem
    audio_dir = file_path.parent / 'Audio'
    output_path = audio_dir / f"{name}.mp3"
    
    audio_dir.mkdir(parents=True, exist_ok=True)
    
    if output_path.exists():
        if file_path.stat().st_mtime < output_path.stat().st_mtime:
            return f"⏩ Skipped (up to date): {file_path.name}"
            
    raw = file_path.read_text(encoding='utf-8')
    clean_text = clean_markdown(raw)
    
    if not clean_text:
        return f"⚠️ Skipped (empty): {file_path.name}"
        
    communicate = edge_tts.Communicate(clean_text, voice)
    await communicate.save(str(output_path))
    
    embed_string = f"![[Audio/{name}.mp3]]"
    if embed_string not in raw:
        match = re.search(r'^---[\s\S]*?---\n', raw, flags=re.MULTILINE)
        if match:
            new_content = raw[:match.end()] + f"{embed_string}\n\n" + raw[match.end():]
        else:
            new_content = f"{embed_string}\n\n{raw}"
            
        file_path.write_text(new_content, encoding='utf-8')
        
    return f"✅ Generated: {output_path.name}"

async def run_with_concurrency(tasks, limit):
    semaphore = asyncio.Semaphore(limit)
    async def sem_task(task):
        async with semaphore:
            return await task()
    
    return await asyncio.gather(*(sem_task(t) for t in tasks))

async def audio_refinery(target: str, voice: str = DEFAULT_VOICE) -> str:
    target_path = Path(target)
    files = []
    
    if not target_path.exists():
        raise RuntimeError(f"Path not found: {target}")
        
    if target_path.is_file() and target_path.suffix in ['.md', '.txt']:
        files.append(target_path)
    elif target_path.is_dir():
        for root, dirs, filenames in os.walk(target_path):
            if 'Audio' in dirs:
                dirs.remove('Audio')
            for f in filenames:
                if f.endswith('.md'):
                    files.append(Path(root) / f)
                    
    if not files:
        return "⚠️ No Markdown files found."
        
    tasks = [lambda f=f: process_file(f, voice) for f in files]
    results = await run_with_concurrency(tasks, CONCURRENCY)
    
    return "\n".join(results)

if __name__ == "__main__":
    if len(sys.argv) < 2:
        print("Usage: python audio_refinery.py <file_or_dir> [voice]", file=sys.stderr)
        sys.exit(1)
        
    target_arg = sys.argv[1]
    voice_arg = sys.argv[2] if len(sys.argv) > 2 else DEFAULT_VOICE
    
    print("🎧 Audio Refinery starting...")
    try:
        output = asyncio.run(audio_refinery(target_arg, voice_arg))
        print(output)
    except Exception as e:
        print(f"Error: {e}", file=sys.stderr)
        sys.exit(1)
